### ISSUE TITLE
`@on` decorator matches widget subclass unexpectedly

### ISSUE THREAD

**Issue by darrenburns** [Textualize/textual#4968]
The code below fires a notification when you click the button.

```python
from __future__ import annotations

from textual import on
from textual.app import App, ComposeResult
from textual.widgets import Button


class MyButton(Button):
    pass


class MyApp(App[None]):
    def compose(self) -> ComposeResult:
        yield Button("Normal Button")

    @on(MyButton.Pressed)
    def notify_on_press_my_button(self) -> None:
        self.notify("MyButton.Pressed matched!")


if __name__ == "__main__":
    MyApp().run()

```

On clicking a `Button`, I would not expect a handler decorated with `@on(MyButton.Pressed)` to fire.

If I click a `MyButton`, I'd expect a `@on(Button.Pressed)` handler to fire, and I've confirmed this works as I expect.

**Comment by willmcgugan** [comment]
@darrenburns The current behavior fits my understanding, but I am not apposed to changing it.

If you subclass `Button`, you don't also subclass `Button.Pressed`

```python
>>> from textual.widgets import Button
>>> class MyButton(Button):
...     pass
...
>>> MyButton.Pressed is Button.Pressed
True
```

So as far as the `on` decorator is concerned, they are the same event.

Feel free to propose a solution...

**Comment by darrenburns** [comment]
Yeah, I understand the mechanics, but if the identity check `is` is how the matching is happening, then I reckon we can do better. It's just such a huge footgun and so unintuitive that I don't think it could be considered anything other than a bug.

If I create a subclass like `class DeleteAllFilesOnMyHardDriveButton(Button)` with a handler like `@on(DeleteAllFilesOnMyHardDriveButton.Pressed)`, I think it's reasonable to expect that would only fire when I click that specific type of button and not the "back up all my files safely" button on the other side of the screen. (Example intended to frighten 😂)

I have some ideas for how we can fix it but will wait until the theme stuff is out of the way.

**Comment by HarperZ9** [comment]
I traced this on current `main` because the failure mode looked like it might be a small `@on` dispatch issue. The current behavior is coming from two separate facts:

- `MyButton.Pressed is Button.Pressed` is `True`, so `@on(MyButton.Pressed)` stores the same message class as `@on(Button.Pressed)`.
- `Button` currently posts `Button.Pressed(self)`, so even if inherited widget message classes were later made owner-specific, the producer side would also need to emit via the runtime widget class, e.g. `type(self).Pressed(self)`, or another equivalent owner-aware path.

I also verified the current workaround on the clone:

```python
@on(Button.Pressed, MyButton)
def selected(self, event: Button.Pressed) -> None:
    ...
```

That only fired for `MyButton` in a two-button test app, while `@on(MyButton.Pressed)` fired for both `Button` and `MyButton`.

So I do not think this can be fixed inside `_on.on()` alone: by the time the decorator receives `message_type`, the subclass owner information has already been lost. A real fix probably needs an explicit design choice, either owner-specific message classes for inherited widget message types plus producer updates, or a new decorator/API form that carries the desired control type separately from the message class.
